File: src/BlazorToast/src/ToastService.ts
interface IToastService {
  IsSupported(): boolean;
  RequestPermission(): Promise<string>;
  Create(title: string, options: object): object;
}

class ToastService implements IToastService {
  public RequestPermission = (): Promise<string> => {
    return new Promise((resolve, reject) => {
      Notification.requestPermission((permission) => {
        resolve(permission);
      });
    });
  }

  public IsSupported = (): boolean => {
    if (("Notification" in window))
      return true;
    return false;
  }

  public Create = (title: string, options: object): object => {
    var note = new Notification(title, options);
    return note;
  }
}

let toastService: ToastService = new ToastService();

export function isSupported(): boolean {
    return toastService.IsSupported();
}

export function requestPermission(): Promise<string> {
    return toastService.RequestPermission();
}

export function create(title: string, options: object): object {
    return toastService.Create(title, options);
}

File: src/BlazorToast/ToastService.cs
using System;
using System.Threading.Tasks;
using Microsoft.JSInterop;

namespace BlazorToast
{
    internal class ToastService : IAsyncDisposable, IToastService
    {
        private const string IsSupportedMethod = "isSupported";
        private const string RequestPermissionMethod = "requestPermission";
        private const string CreateMethod = "create";

        private readonly Lazy<Task<IJSObjectReference>> moduleTask;

        public ToastService(IJSRuntime jsRuntime)
        {
            moduleTask = new(() => jsRuntime.InvokeAsync<IJSObjectReference>(
               "import", "./_content/BlazorToast/ToastService.js").AsTask());
        }

        public async ValueTask<bool> IsSupported()
        {
            var module = await moduleTask.Value;

            return await module.InvokeAsync<bool>(IsSupportedMethod);
        }

        public async ValueTask<NotificationPermission> RequestPermission()
        {
            var module = await moduleTask.Value;

            var invocationResult = await module.InvokeAsync<string>(RequestPermissionMethod);

            if (!Enum.TryParse(invocationResult, out NotificationPermission permission))
            {
                throw new InvalidOperationException($"Unknown permission value: '{invocationResult}'");
            }

            return permission;
        }

        public async ValueTask<object> Create(string title, object options)
        {
            var module = await moduleTask.Value;

            return await module.InvokeAsync<object>(CreateMethod, title, options);
        }

        public async ValueTask DisposeAsync()
        {
            if (moduleTask.IsValueCreated)
            {
                var module = await moduleTask.Value;
                await module.DisposeAsync();
            }
        }
    }
}
